src/software/ai/navigator/path_planner/theta_star_path_planner.cpp

```diff
@@ -55,31 +77,121 @@ double ThetaStarPathPlanner::coordDistance(const Coordinate &coord1,
     return distance(p1, p2);
 }
 
-bool ThetaStarPathPlanner::lineOfSight(const Coordinate &coord1, const Coordinate &coord2)
+bool ThetaStarPathPlanner::lineOfSight(const Coordinate &coord0, const Coordinate &coord1)
 {
-    CoordinatePair coord_pair(coord1, coord2);
-    // If we haven't checked this Coordinate pair for intersects before, check it now
-    auto line_of_sight_cache_it = line_of_sight_cache.find(coord_pair);
-    if (line_of_sight_cache_it == line_of_sight_cache.end())
+    unsigned int x0 = coord0.row();
+    unsigned int y0 = coord0.col();
+    unsigned int x1 = coord1.row();
+    unsigned int y1 = coord1.col();
+    std::vector<Coordinate> line_of_sight_coords;
+
+    int dy = y1 - y0;
+    int dx = x1 - x0;
+
+    if (std::abs(dy) < std::abs(dx))
+    {
+        if (x0 > x1)
+        {
+            return checkLineLow(coord1, coord0);
+        }
+        else
+        {
+            return checkLineLow(coord0, coord1);
+        }
+    }
+    else
+    {
+        if (y0 > y1)
+        {
+            return checkLineHigh(coord1, coord0);
+        }
+        else
+        {
+            return checkLineHigh(coord0, coord1);
+        }
+    }
+}
+
+bool ThetaStarPathPlanner::checkLineLow(const Coordinate &coord0,
+                                        const Coordinate &coord1)
+{
+    unsigned int x0 = coord0.row();
+    unsigned int y0 = coord0.col();
+    unsigned int x1 = coord1.row();
+    unsigned int y1 = coord1.col();
+
+    int dx = x1 - x0;
+    int dy = y1 - y0;
+    int yi = 1;
+
+    if (dy < 0)
     {
-        Segment seg(convertCoordToPoint(coord1), convertCoordToPoint(coord2));
-        bool has_line_of_sight = true;
+        yi = -1;
+        dy = -dy;
+    }
+    int D = (2 * dy) - dx;
+    int y = y0;
 
-        for (const auto &obstacle : obstacles)
+    for (unsigned int x = x0; x <= x1; x++)
+    {
+        if (isBlocked(Coordinate(x, y)))
         {
-            if (obstacle->intersects(seg))
-            {
-                has_line_of_sight = false;
-                break;
-            }
+            // No line of sight since a coordinate in the path from coord0 to coord1 is
+            // blocked
+            return false;
+        }
+        if (D > 0)
+        {
+            y += yi;
+            D += 2 * (dy - dx);
+        }
+        else
+        {
+            D += 2 * dy;
         }
+    }
+    return true;
+}
 
-        // We use the opposite convention to indicate blocked or not
-        line_of_sight_cache[coord_pair] = has_line_of_sight;
-        return has_line_of_sight;
+bool ThetaStarPathPlanner::checkLineHigh(const ThetaStarPathPlanner::Coordinate &coord0,
```

Comment:
looks like quite a bit of duplicated code between this function and checkLineLow, can we refactor it, e.g. pass a flag high or low into a checkLine function

Reply:
I agree, however I think it would make the methods much less readable if they are combined. Though I can combine them and then not refer to `x` and `y` and instead refer to (eg.) `main_axis` and `secondary_axis`.